### PR TITLE
New article: git commit partially verified

### PR DESCRIPTION
**Pull request by josecelano**

### CODE REVIEW

**Comment by josecelano** [comment]
I've changed it back to draft because I want to make some minor changes/notes.

- The "partially verified" status appears when you use the console to merge with your committer info.
- If you, for example, merge the PR using the GitHub UI, the commit appears "verified". The committer is the GitHub bot and the author is the logged-in user. [Sample commit](https://github.com/Nautilus-Cyberneering/GPG-Bootcamp/commit/c031a4a2d163efd59cd78339ac19543e525c5fbb).

I have to make more explicit the differences between using the GitHub UI and the console. And for "merge" and "rebase and merge" cases.

**Comment by josecelano** [comment]
> I've changed it back to draft because I want to make some minor changes/notes.
> 
> * The "partially verified" status appears when you use the console to merge with your committer info.
> * If you, for example, merge the PR using the GitHub UI, the commit appears "verified". The committer is the GitHub bot and the author is the logged-in user. [Sample commit](https://github.com/Nautilus-Cyberneering/GPG-Bootcamp/commit/c031a4a2d163efd59cd78339ac19543e525c5fbb).
> 
> I have to make more explicit the differences between using the GitHub UI and the console. And for "merge" and "rebase and merge" cases.

I'm using [this repo](https://github.com/josecelano-test/git-commit-signature) for the experiments.

**Comment by da2ce7** [comment]
@josecelano Great investigation!

You seem to be mixing up Committer, Author, and the [Commit Signatures](https://git-scm.com/docs/signature-format/en#_commit_signatures)

Git doesn't have a committer, or author signature, it simply has a Commit Signature.  - GitHub has the concept of a committer, author, and commit signature relationship.

Either the commit signature's matches key uploaded to the github-profile of the committer + author, only committer, or none.

**Comment by josecelano** [comment]
> @josecelano Great investigation!
> 
> You seem to be mixing up Committer, Author, and the [Commit Signatures](https://git-scm.com/docs/signature-format/en#_commit_signatures)
> 
> Git doesn't have a committer, or author signature, it simply has a Commit Signature. - GitHub has the concept of a committer, author, and commit signature relationship.
> 
> Either the commit signature's matches key uploaded to the github-profile of the committer + author, only committer, or none.

Hey, @da2ce7 thank you for clarifying it. I'm still trying to understand the differences between Git and Github and that's a really good point. I will polish the article as I get a better understanding. I tend to assign the signature to the committer info because in the Git global configuration the signing key is in the same section as the user info. In my case:

```
$ git config -l | grep user
user.name=Jose Celano
user.email=josecelano@gmail.com
user.signingkey=58508C7950C7B7A2
```

I supposed the "user" attributes are the "committer attributes". Maybe that option should be in the "commit" section next to the "gpgsign" option.

On the Git official documentation you can read:

```
-S[<keyid>]
--gpg-sign[=<keyid>]
--no-gpg-sign
GPG-sign commits. The keyid argument is optional and defaults to the committer identity; if specified, it must be stuck to the option without a space. --no-gpg-sign is useful to countermand both commit.gpgSign configuration variable, and earlier --gpg-sign.
```

From: https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---gpg-signltkeyidgt

I understand you can sign the commit with any other GPG key not related to the committer, even though I can't imagine a use case now for it. Well, maybe you could give the commit object to the author so the author can sign it and then they can give the detached signature back to the committer so they can include it in the commit header. Maybe GitHub should include a "sign request" :-).

In fact, in the more specific [documentation about commit signature](https://git-scm.com/docs/signature-format#_commit_signatures), they do not use the same identity for the committer and the signer.

I have to correct that in the article.

**Comment by josecelano** [comment]
hi @da2ce7 in this [PR](https://github.com/Nautilus-Cyberneering/git-queue/commit/aa9089cdb0646bc1efabb2c0c0c81ee11038827c) in the [git-queue repo](https://github.com/Nautilus-Cyberneering/git-queue) I merged an @ivanramosnet commit with the bot account (git identity and GPG key).

We do not have the partially verified message from GitHub and this is the commit info:

```s
$ git show --pretty=fuller --show-signature aa9089cdb0646bc1efabb2c0c0c81ee11038827c
commit aa9089cdb0646bc1efabb2c0c0c81ee11038827c
gpg: Signature made mar 10 may 2022 19:08:17 WEST
gpg:                using RSA key 76BF41FD8C0589E1D5288B29E7279F2518CA55C6
gpg: Good signature from "Nautilus Cyberneering [bot] (Online Key) <bot@nautilus-cyberneering.de>" [ultimate]
Author:     Iván Ramos Jiménez <info@ivan.ramos.name>
AuthorDate: Tue May 10 18:00:59 2022 +0100
Commit:     Nautilus Cyberneering [bot] [josecelano] <bot@nautilus-cyberneering.de>
CommitDate: Tue May 10 19:08:13 2022 +0100

    refactor: remove unused InvalidShortHashError error

```

Another partially verified example is [the previous commit](https://github.com/Nautilus-Cyberneering/git-queue/commit/d47bd36bdad81487862ae73f3464a89c1a4fe9bd).

```s
$ git show --pretty=fuller --show-signature d47bd36bdad81487862ae73f3464a89c1a4fe9bd
commit d47bd36bdad81487862ae73f3464a89c1a4fe9bd
gpg: Signature made mar 26 abr 2022 10:32:32 WEST
gpg:                using RSA key 76BF41FD8C0589E1D5288B29E7279F2518CA55C6
gpg: Good signature from "Nautilus Cyberneering [bot] (Online Key) <bot@nautilus-cyberneering.de>" [ultimate]
Author:     Jose Celano <josecelano@gmail.com>
AuthorDate: Tue Apr 26 10:17:45 2022 +0100
Commit:     Nautilus Cyberneering [bot] [josecelano] <bot@nautilus-cyberneering.de>
CommitDate: Tue Apr 26 10:32:30 2022 +0100

    fix: [#173] update node dependencies

```

image

I do not see any difference. I do not know why one is verified and the other one is partially verified. I have to find it out. Anyway, I have to finish running all cases in the [test repo](https://github.com/josecelano-test/git-commit-signature/issues?q=is%3Aissue+is%3Aclosed).

**Comment by josecelano** [comment]
This is the way I'm merging PR into `main` branch:

Update `main`:

```s
git checkout main
git pull
```
Checkout branch in the fork:

```s
git checkout -b ivanramosnet-issue-196-remove-deprecated-comment main
git pull https://github.com/ivanramosnet/git-queue.git issue-196-remove-deprecated-comment
```

Merge branch in the fork with a FF merge:

```s
git checkout main
gi merge --ff-only ivanramosnet-issue-196-remove-deprecated-comment
git show --pretty=fuller --show-signature HEAD
git push origin main
```

**Comment by josecelano** [comment]
> hi @da2ce7 in this [PR](https://github.com/Nautilus-Cyberneering/git-queue/commit/aa9089cdb0646bc1efabb2c0c0c81ee11038827c) in the [git-queue repo](https://github.com/Nautilus-Cyberneering/git-queue) I merged an @ivanramosnet commit with the bot account (git identity and GPG key).
> 
> We do not have the partially verified message from GitHub and this is the commit info:
> 
> ```m68k
> $ git show --pretty=fuller --show-signature aa9089cdb0646bc1efabb2c0c0c81ee11038827c
> commit aa9089cdb0646bc1efabb2c0c0c81ee11038827c
> gpg: Signature made mar 10 may 2022 19:08:17 WEST
> gpg:                using RSA key 76BF41FD8C0589E1D5288B29E7279F2518CA55C6
> gpg: Good signature from "Nautilus Cyberneering [bot] (Online Key) <bot@nautilus-cyberneering.de>" [ultimate]
> Author:     Iván Ramos Jiménez <info@ivan.ramos.name>
> AuthorDate: Tue May 10 18:00:59 2022 +0100
> Commit:     Nautilus Cyberneering [bot] [josecelano] <bot@nautilus-cyberneering.de>
> CommitDate: Tue May 10 19:08:13 2022 +0100
> 
>     refactor: remove unused InvalidShortHashError error
> ```
> 
> Another partially verified example is [the previous commit](https://github.com/Nautilus-Cyberneering/git-queue/commit/d47bd36bdad81487862ae73f3464a89c1a4fe9bd).
> 
> ```m68k
> $ git show --pretty=fuller --show-signature d47bd36bdad81487862ae73f3464a89c1a4fe9bd
> commit d47bd36bdad81487862ae73f3464a89c1a4fe9bd
> gpg: Signature made mar 26 abr 2022 10:32:32 WEST
> gpg:                using RSA key 76BF41FD8C0589E1D5288B29E7279F2518CA55C6
> gpg: Good signature from "Nautilus Cyberneering [bot] (Online Key) <bot@nautilus-cyberneering.de>" [ultimate]
> Author:     Jose Celano <josecelano@gmail.com>
> AuthorDate: Tue Apr 26 10:17:45 2022 +0100
> Commit:     Nautilus Cyberneering [bot] [josecelano] <bot@nautilus-cyberneering.de>
> CommitDate: Tue Apr 26 10:32:30 2022 +0100
> 
>     fix: [#173] update node dependencies
> ```
> 
> image
> 
> I do not see any difference. I do not know why one is verified and the other one is partially verified. I have to find it out. Anyway, I have to finish running all cases in the [test repo](https://github.com/josecelano-test/git-commit-signature/issues?q=is%3Aissue+is%3Aclosed).

I suppose the commit is verified because @ivanramosnet has not enabled the "[vigilant mode](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#statuses-with-vigilant-mode-enabled)"

image

I forgot you have no enable it in your profile.

**Comment by josecelano** [comment]
> @josecelano Great investigation!
> 
> You seem to be mixing up Committer, Author, and the [Commit Signatures](https://git-scm.com/docs/signature-format/en#_commit_signatures)
> 
> Git doesn't have a committer, or author signature, it simply has a Commit Signature. - GitHub has the concept of a committer, author, and commit signature relationship.
> 
> Either the commit signature's matches key uploaded to the github-profile of the committer + author, only committer, or none.

hi @da2ce7 I've reviewed the article and rewritten some parts to include your clarification.

**Comment by da2ce7** [comment]
@josecelano Good Clarification. 👍

File: docs/010_GPG-Git-commits-partially-verified.md [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+![Commit with partially verified signature](../media/010/commit-with-partially-verified-signature-on-github.png)
+
+Maybe losing the author's signature is not an issue from the package user or maintainers' point of view. However, from the author's point of view, you can not distinguish anymore between your legitimate commits and the fake ones. Someone else could pretend you have collaborated with their project. It is surprisingly easy to do that. You only need to create a commit and override the author with any users on GitHub.
+
+## Conclusion
+
+When you use signed commits:
+
+- Be aware of the Git limitation of one single signature per commit.
+- Be also aware of what happens to signatures when you use Git commands like "merge" and "rebase".
```
Comment (cgbosse) [inline docs/010_GPG-Git-commits-partially-verified.md:426]: Possible new article name: Git Merge and rebase: What happens under the hood

File: docs/010_GPG-Git-commits-partially-verified.md [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+```text
+GIT_AUTHOR_NAME
+GIT_AUTHOR_EMAIL
+GIT_COMMITTER_NAME
+GIT_COMMITTER_EMAIL
+```
+
+For the author you have a command option too (`git commit -m "..." --author="Your Name <your@email.com>"`).
+
+You can sign commits using the committer GPG key. Git does not support adding the author signature in addition to the committer signature.
```
[the remark this answers is not in the document]
Comment (josecelano) [reply docs/010_GPG-Git-commits-partially-verified.md:33]: In fact, I had to remove that line. In Git the commit signature does not have to be done with the committer GPG key.

File: docs/010_GPG-Git-commits-partially-verified.md [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+
+We have not found a reason why Git does not support multiple signatures. Since Git only relays on the GPG console command, we thought the reason was that GPG did not support it either, but as you can read on this [Stackoverflow answer](https://stackoverflow.com/questions/37725969/several-pgp-signatures-for-one-file/37759313#37759313), there are two ways to add multiple signatures to a file using GPG.
+
+The first solution is to sign the contents with one key and sign the result again with the following key until you sign it with all the keys.
+
+The second solution would be using "detached signatures", which means the signature is not included in the signed document. We suppose Git could have used this approach, and [this article from Owen Jacobson(](https://grimoire.ca/git/detached-sigs/) describes it in detail. But it is not the case. Although Git uses "detached signatures", the signature is included as a commit header inside the "full" commit object that is used to calculate the commit hash. That means changing the signature changes the commit hash.
+
+## Git merge, rebase and GPG signature
```
[the remark this answers is not in the document]
Comment (josecelano) [reply docs/010_GPG-Git-commits-partially-verified.md:121]: We can add a new idea for a new article [here](https://github.com/Nautilus-Cyberneering/GPG-Bootcamp/discussions/categories/ideas). I've added a link to the other article where Owen explains how GIt could have been implemented. There is [another article](https://mikegerwitz.com/2012/05/a-git-horror-story-repository-integrity-with-signed-commits#trust-ensure) that explains how you can implement it using the current Git version.

Maybe we should also add an index of articles with a short description to the repo, not only at the end of each article.

File: docs/010_GPG-Git-commits-partially-verified.md (hunk 1 above)
Comment (josecelano) [reply docs/010_GPG-Git-commits-partially-verified.md:426]: OK, but It should be focused on how the command affects the commit signatures.

**Comment by josecelano** [comment]
hi @cgbosse I applied your fixes/suggestions.

**Pull request merged**